rc/extra/protobuf.kak

```diff
@@ -0,0 +1,89 @@
+# https://developers.google.com/protocol-buffers/
+
+# Detection
+# ‾‾‾‾‾‾‾‾‾
+
+hook global BufCreate .*\.proto$ %{
+    set-option buffer filetype protobuf
+}
+
+# Highlighters
+# ‾‾‾‾‾‾‾‾‾‾‾‾
+
+add-highlighter shared/protobuf regions
+add-highlighter shared/protobuf/code default-region group
+add-highlighter shared/protobuf/double_string region '"' (?<!\\)(\\\\)*" fill string
+add-highlighter shared/protobuf/comment region /\* \*/ fill comment
+add-highlighter shared/protobuf/comment_line region '//' $ fill comment
+
+add-highlighter shared/protobuf/code/ regex %{(0x)?[0-9]+\b} 0:value
+
+evaluate-commands %sh{
+    # Grammer
+    keywords="default|deprecated|enum|extend|import|message|oneof|option"
+    keywords="${keywords}|package|service|syntax"
+    attributes="optional|repeated|required"
+    types="double|float|int32|int64|uint32|uint64|sint32|sint64|fixed32|fixed64"
+    types="${types}|sfixed32|sfixed64|bool|string|bytes|rpc"
+    values="false|true"
+
+    # Add the language's grammer to the static completion list
+    printf '%s\n' "hook global WinSetOption filetype=protobuf %{
+        set-option window static_words ${keywords} ${attributes} ${types} ${values}
+    }" | tr '|' ' '
+
+    # Highlight keywords
+    printf %s "
+        add-highlighter shared/protobuf/code/keywords regex \b(${keywords})\b 0:keyword
+        add-highlighter shared/protobuf/code/attributes regex \b(${attributes})\b 0:attribute
+        add-highlighter shared/protobuf/code/types regex \b(${types})\b 0:type
+        add-highlighter shared/protobuf/code/values regex \b(${values})\b 0:value
+    "
+}
+
+# Commands
+# ‾‾‾‾‾‾‾‾
+
+
+define-command -hidden protobuf-indent-on-newline %~
+    evaluate-commands -draft -itersel %[
+        # preserve previous line indent
+        try %{ execute-keys -draft \;K<a-&> }
+        # indent after lines ending with {
+        try %[ execute-keys -draft k<a-x> <a-k> \{\h*$ <ret> j<a-gt> ]
+        # cleanup trailing white spaces on the previous line
+        try %{ execute-keys -draft k<a-x> s \h+$ <ret>d }
+        # copy // comments prefix
+        try %{ execute-keys -draft \;<c-s>k<a-x> s ^\h*\K/{2,}(\h*(?=\S))? <ret> y<c-o>P<esc> }
+    ]
+~
+
+define-command -hidden protobuf-indent-on-opening-curly-brace %[
+    # align indent with opening paren when { is entered on a new line after the closing paren
+    try %[ execute-keys -draft -itersel h<a-F>)M <a-k> \A\(.*\)\h*\n\h*\{\z <ret> s \A|.\z <ret> 1<a-&> ]
+]
+
+define-command -hidden protobuf-indent-on-closing-curly-brace %[
+    # align to opening curly brace when alone on a line
+    try %[ execute-keys -itersel -draft <a-h><a-k>^\h+\}$<ret>hms\A|.\z<ret>1<a-&> ]
+]
+
+# Initialization
+# ‾‾‾‾‾‾‾‾‾‾‾‾‾‾
+
+hook -group protobuf-highlight global WinSetOption filetype=protobuf %{ add-highlighter window/protobuf ref protobuf }
+
+hook global WinSetOption filetype=protobuf %[
+    # cleanup trailing whitespaces when exiting insert mode
```

Comment:
Are trailing whitespace strictly forbidden by the protobuf syntax? If not, then this hook should be removed, as it's not the role of syntax support scripts to decide how the buffer should be formatted.

Reply:
It's not strictly forbidden by c/c++ syntax but the c-family syntax files still strip trailing whitespace on empty lines.

Reply:
I know some language support scripts do it, but that is a mistake that was fixed in other scripts, while it stayed in the older ones.

One example why that hook is annoying from a practical point of view is when you want to paste something: <kbd>i</kbd> `int<space>nvar<space>=<space>` <kbd>esc</kbd><kbd>p</kbd> would remove the space after the equal sign.

Reply:
I raised this as well in IRC and #2175, I think the consensus is that no syntax file should ever touch whitespaces, this should be the job of formatters.

Reply:
That sounds like a sensible change, but if I did that now for the protobuf syntax then it's behavior would be inconsistent from the other syntax files until that change gets made.

Reply:
Yep, but that would be the problem of the other syntax files, not protobuf syntax. In other words, if someone is annoyed by inconsistency they would submit a PR for the other syntax files.

Reply:
Should the cleanup of whitespace following a newline also be removed from the syntax files?

Reply:
Also I feel like something like that should be done all at once for every syntax file with that behavior if this change is going to be made to avoid inconsistency.

Reply:
> Should the cleanup of whitespace following a newline also be removed from the syntax files?

In my mind we should only keep the hooks that help you _type_, things like increase indent after typing `{` and decrease indent after typing `}`. But there should be no hooks that post factum reformat my code — if I added whitespaces after newline, so let them be.

> Also I feel like something like that should be done all at once for every syntax file with that behavior if this change is going to be made to avoid inconsistency.

I will leave this to you and others to decide, I already expressed my opinion 🙂